app/src/main/kotlin/compiler/ast/Function.kt

```diff
@@ -2,13 +2,7 @@ package compiler.ast
 
 import compiler.input.LocationRange
 
-data class Function(
-    override val name: String,
-    val parameters: List<Parameter>,
-    val returnType: Type,
-    val implementation: Implementation,
-    override val location: LocationRange? = null,
-) : NamedNode, AstNode, VariableOwner {
+sealed class MetaFunction : NamedNode, AstNode, VariableOwner {
```

Comment:
Wouldn't it be easier to just add `isGenerator: Boolean` to `Function`? The field `returnType` could then mean either "return type" or "yield type" (which can be thought of as "generalized return"). I think functions and generator functions will actually have quite a lot in common, and this separation may lead to large amount of code duplication.

Reply:
Okay, maybe there won't be too much code duplication if we have this `MetaFunction` class, but still, since `Function` and `Generator` are basically the same thing, the difference being only the type and the interpretation of `returnType`, instead of writing `f is Generator` we may as well write `f.isGenerator`. This would be more consistent with the fact that foreign and local functions have the same type, despite being somewhat different things.

Reply:
Thought that using `MetaFunction` would be more convenient, because it could help find `when` clauses where we'll be missing generator handling, but I agree that it would have a lot in common with normal `Function`s and we should be consistent with `Foreign` functions. Refactored as you suggested.